netty/src/test/java/io/grpc/transport/netty/NettyClientTransportTest.java

```diff
@@ -157,15 +163,61 @@ public void creatingMultipleTlsTransportsShouldSucceed() throws Exception {
     }
   }
 
+  @Test
+  public void bufferedStreamsShouldBeClosedWhenConnectionTerminates() throws Exception {
+    // Create the protocol negotiator.
+    File clientCert = TestUtils.loadCert("ca.pem");
+    SslContext clientContext = GrpcSslContexts.forClient().trustManager(clientCert).build();
+    ProtocolNegotiator negotiator = ProtocolNegotiators.tls(clientContext, address);
+
+    NettyClientTransport transport = new NettyClientTransport(address, NioSocketChannel.class,
+            group, negotiator, DEFAULT_WINDOW_SIZE, DEFAULT_WINDOW_SIZE);
+    transports.add(transport);
+    transport.start(clientTransportListener);
+
+    // Wait for the settings to be propagated to the client so that it will respect
+    // SETTINGS_MAX_CONCURRENT_STREAMS.
+    Thread.sleep(1000);
+
+    // Create 3 streams. The transport will buffer the second and third.
+    TestClientStreamListener[] listeners = new TestClientStreamListener[] {
+        new TestClientStreamListener(), new TestClientStreamListener(),
+        new TestClientStreamListener() };
+    for (TestClientStreamListener listener : listeners) {
+      ClientStream stream = transport.newStream(METHOD, new Metadata.Headers(), listener);
+      stream.request(1);
+      stream.writeMessage(messageStream());
+      stream.flush();
+    }
+    // Wait for the response for the stream that was actually created.
+    listeners[0].responseFuture.get(10, TimeUnit.SECONDS);
+
+    // Now forcibly terminate the connection from the server side.
+    TestSocketChannel socketChannel =
+            (TestSocketChannel) serverListener.transports.get(0).channel();
+    socketChannel.forceShutdown();
+
+    // Now wait for both listeners to be closed.
+    for (TestClientStreamListener listener : listeners) {
+      try {
+        listener.closedFuture.get(10, TimeUnit.SECONDS);
+      } catch (ExecutionException e) {
+        // Expected.
+      }
+    }
+  }
+
   private static InputStream messageStream() {
     return new ByteArrayInputStream(MESSAGE.getBytes());
   }
 
   private static class TestClientStreamListener implements ClientStreamListener {
```

Comment:
Should this be a mock? It seems like verifying with timeouts would do everything this class provides.